### DISCUSSION TITLE
Docker swarm

### DISCUSSION THREAD

**Discussion by suchoss** [canarytrace/documentation#125] [Q&A] [answered]
Hi,

We do not have a k8s environment and I would like to run canary trace in docker swarm.

To be able to do that correctly I need to know a few things:
1) Why do a chrome container needs `dshm` memory volume?
2) Why do a canarytrace container needs `assets` volume?

If I take in a consideration the fact that the canary trace is being run once in a while, then I do not understand this persistence.

How do the canarytrace image communicate with the chrome image? Do they use only port 4444 on a localhost? Is it possible to use some ENV to change the address of the chrome?

Thanks

**Comment by rdpanek** [comment, marked as answer]
Hello @suchoss 👋

- `dsh` ☝️ When executing docker run for an image that contains a browser please use the flag --shm-size=2g to use the host's shared memory. Reason is known bug of Chrome https://bugs.chromium.org/p/chromium/issues/detail?id=522853
- `assets` directory is for statics files generated by Canarytrace, e.g. trace reporter. https://canarytrace.com/docs/features/lighthouse

Second docker image contains chromedriver and Canarytrace use default port 4444 for sending commands and 9222 for remote debugging of the browser. There is no reason to change these default ports of the browser.

Thanks

**Reply by suchoss** [reply]
Hello @rdpanek 

the first part (`dsh` & `assets`) is clear 👌 

The second part - 🤔 I have no issue with ports (even though I might have). I have an issue with localhost reference when I want to run it outside of k8s. That is why I asked if there is an environment variable where I can tell to Canarytrace "Hey, your chrome is placed somewhere else." 👈  👉 

I guess I can find a workaround (although a bit painful 😢) for that in Docker swarm. I am just thinking that this would help people and companies which are not using k8s.

Also I didn't know about 9222 - thanks for mentioning it.

**Reply by rdpanek** [reply]
We released Canarytrace with best practices for Ops and Tests and one of BP is used Kubernetes :D :D :D It means, minumum maintenance for operators/users. So, configuration for browser is hardcoded in WDIO (WebdriverIO) configuration:

```
exports.config = {
  username: username,
  password: password,
  hostname: 'localhost',
  port: 4444,
  path: '/',
  capabilities: [{
    browserName: 'chrome',
    acceptInsecureCerts: true,
    'goog:chromeOptions': {
      args: [
        '--remote-debugging-port=9222',
        '--remote-debugging-host=0.0.0.0',
        '--start-fullscreen',
        '--disable-notifications',
        '--window-size=1920,1080',
        '--no-default-browser-check',
        '--no-first-run',
        '--disable-breakpad',
        '--disable-component-update',
        '--disable-sync',
        '--allow-running-insecure-content',
        '--disable-client-side-phishing-detection',
        '--disable-default-apps',
        '--disable-device-discovery-notifications',
        '--disable-domain-reliability',
        '--disable-background-timer-throttling',
        '--load-media-router-component-extension=0',
        '--mute-audio',
        '--disable-hang-monitor',
        '--password-store=basic',
        '--dont-require-litepage-redirect-infobar',
        '--override-https-image-compression-infobar',
        '--disable-site-isolation-trials'
      ]
    }
  }],
```
But for development we use ENVs `export HOSTNAME=localhost` and `export PORT=4444` please try it.

**Reply by suchoss** [reply]
Will try. Thanks
